swift/ql/lib/codeql/swift/security/PathInjection.qll

```diff
@@ -0,0 +1,99 @@
+/** Provides classes and predicates to reason about path injection vulnerabilities. */
+
+import swift
+private import codeql.swift.dataflow.DataFlow
+private import codeql.swift.dataflow.ExternalFlow
+
+/** A data flow sink for path injection vulnerabilities. */
+abstract class PathInjectionSink extends DataFlow::Node { }
+
+/** A sanitizer for path injection vulnerabilities. */
+abstract class PathInjectionSanitizer extends DataFlow::Node { }
+
+/**
+ * A unit class for adding additional taint steps.
+ *
+ * Extend this class to add additional taint steps that should apply to paths related to
+ * path injection vulnerabilities.
+ */
+class PathInjectionAdditionalTaintStep extends Unit {
+  /**
+   * Holds if the step from `node1` to `node2` should be considered a taint
+   * step for paths related to path injection vulnerabilities.
+   */
+  abstract predicate step(DataFlow::Node node1, DataFlow::Node node2);
+}
+
+private class DefaultPathInjectionSink extends PathInjectionSink {
+  DefaultPathInjectionSink() { sinkNode(this, "path-injection") }
+}
+
+private class DefaultPathInjectionSanitizer extends PathInjectionSanitizer {
+  DefaultPathInjectionSanitizer() { none() } // TODO: Implement a proper path sanitizer
```

Comment:
A proper path sanitizer can become quite complex (see [the Java implementation](https://github.com/github/codeql/blob/7456f3750d36f56d413839636513849c980b5c97/java/ql/lib/semmle/code/java/security/PathSanitizer.qll) for a reference).

I think it would make review easier if we created that in its own PR. But let me know if you disagree and want to see that in this PR (as of now, the recommendation made in the QHelp (`PathInjectionGood.swift` sample) would not actually make alerts disappear, which is always a bad user experience).

Reply:
What kind of level of false positives are we likely to see with no sanitizer here?  Right now lack of results is a much higher priority for Swift than false positives (within reason) and I've deferred other similar work until well into 2023 for this reason.

On the flip side its nice to get things finished while they're on your mind.  And I agree that giving advice in the `qhelp` that won't make the alert go away is a problem.

Reply:
> What kind of level of false positives are we likely to see with no sanitizer here?

I'd need to run the query on a decent amount of projects to be sure, but intuitively, any app doing filesystem operations based on user input would get flagged, regardless of how well they validate the input. So it depends on how many apps do the risky thing (user input into filesystem sink), and how many of those sanitize/validate well.

> On the flip side its nice to get things finished while they're on your mind. And I agree that giving advice in the qhelp that won't make the alert go away is a problem.

As a compromise, we could write a very basic sanitizer for the moment that does the bare minimum to detect at least the simplest sanitization/validation logic (including the GOOD case in the `qhelp`), and defer the rest until we have a proper Guards library.

Reply:
Sounds like a plan. :+1:

Reply:
Done in 52ebf66. It took a while because I was banging my head against the wall until I figured out https://github.com/github/codeql/pull/11469 was needed.